### PR TITLE
[PS-2251] Implement argon2 kdf

### PR DESCRIPTION
**Pull request by quexten**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This pull request implements argon2id support via wasm (and native bindings on node js).
Some points of consideration: I tested on desktop (linux), web (chrome, firefox), manifest-v2 browser (chrome), manifest-v3 browser (chrome), cli (linux). All other platforms are untested by me, any help testing is appreciated, especially on safari and edge.

For now only memory is configurable, but in a future pull request me might introduce a kdfOptions object, to expose more configuration options (iterations, parallelism) to the user.

## Code changes

- manifestv3.json: csp should be "extension page*s*", and add wasm-unsafe-eval so we can load the wasm. (Is there a better way here? I tried configuring the CSP via the hash of the WASM file but wasn't able to get it to work. On the other hand manifest.json alread has "unsafe-eval" set which is way more loose (a superset of wasm-unsafe-eval), so it might not be an issue?
- webpack.config.js: Add rules required for wasm, and fallbacks for fs, path required by the argon2 module
- webpack.main.js: Add copy step for the wasm module for the desktop app
- webpack.render.js: Add wasm loader configurations
- package.json: Add argon2 browser module
- change-kdf.component.html: Add the UI for argon2 configuration
- change-kdf-component.ts: Add argon2 configuration code
- messages.json: Add internationalized texts for agno2 configuration UI
- cryptoFunction.service.ts: Add argon2 interface
- kdfType.ts: Add argon2 as type with id 2 (id 1 is taken by scrypt in another pull request)
- crypto.service.ts: Add the case for argon2, iterations and parallelization are fixed for now
- webCryptoFunction.service.ts, node-crypto-function.service.ts: Implement argon2 calls, using WASM in the former service and bindings in the latter service. We could also use WASM in node-crypto-function.service.ts but this would likely be slower than the native bindings.

## Screenshots
screenshot

### CODE REVIEW

**Comment by quexten** [comment]
Currently the pull request only stores memory (in the kdfIterations value).
More clean would be to directly store the memory in a dedicated field, and use the kdfIterations for argon2's iterations. Here is a pull request enabling that functionality on the back-end side. If it is accepted I will update this pull-request to take advantage of the new API..

https://github.com/bitwarden/server/pull/2583

**Comment by quexten** [comment]
> Currently the pull request only stores memory (in the kdfIterations value). More clean would be to directly store the memory in a dedicated field, and use the kdfIterations for argon2's iterations. Here is a pull request enabling that functionality on the back-end side. If it is accepted I will update this pull-request to take advantage of the new API..
> 
> [bitwarden/server#2583](https://github.com/bitwarden/server/pull/2583)

The server implementation still needs some work, but in the meantime the client has been updated to support more options.
It required replacing a lot of references to kdfIterations by a "KdfConfig" object. The kdfMemory and kdfParallelism fields are always optional.

Screenshot of the the configuration:
screenshot

**Comment by quexten** [comment]
Since in the mobile pull request we are doing the kdf support first with fixed memory and parallelism, should I split this pull request into 2 requests, one doing the kdf support (same as on mobile), and one adding the kdf configuration (since that relies on server changes anyways)?

**Comment by quexten** [comment]
> > Since in the mobile pull request we are doing the kdf support first with fixed memory and parallelism, should I split this pull request into 2 requests, one doing the kdf support (same as on mobile), and one adding the kdf configuration (since that relies on server changes anyways)?
> 
> Yes, let's do that. I think we will probably want to deploy the client support ahead of the configuration anyways so that people can get client updates propagated out ahead of being able to enable it.

Okay, pulled out all the kdfConfiguration related changes and rebased onto latest master.

**Comment by kspearrin** [comment]
@Hinton Can you look into the wasm implementation. Is this how we want to do it?

**Comment by quexten** [comment]
Had a wrong commit from another branch on my local master, so had to force push one more time.

**Comment by quexten** [comment]
> @Hinton Can you look into the wasm implementation. Is this how we want to do it?

By the way about the wasm implementation: The argon2-browser project states that on node.js, native bindings should be used(https://www.npmjs.com/package/argon2), which I did. WASM would also work in node.js though. We could save an implementation/dependency by only using the WASM implmentation if we don't need the extra speed since as long as it is acceptable in WASM in the browser, it should be acceptable for the CLI too speed wise, right?

File: apps/web/src/app/settings/change-kdf.component.html [hunk 1]
```diff
@@ -39,34 +39,66 @@ <h1>{{ "encKeySettings" | i18n }}</h1>
...
+          <ng-container *ngIf="kdf == 0">
+            <p>{{ "kdfIterationsDesc" | i18n: (recommendedKdfIterations | number) }}</p>
+            <strong>{{ "warning" | i18n }}</strong
+            >: {{ "kdfIterationsWarning" | i18n: (50000 | number) }}
+          </ng-container>
+          <ng-container *ngIf="kdf == 2">
+            <p>
+              {{ "argon2MemoryDesc" | i18n: (recommendedMemory | number) }}
+            </p>
+            <bit-callout type="warning"> {{ "argon2Warning" | i18n }}} </bit-callout>
```
[the remark this answers is not in the document]
Comment (quexten) [reply apps/web/src/app/settings/change-kdf.component.html:107]: Yeah, fair.[ Only IE](https://caniuse.com/wasm) really, but I'm not even sure bitwarden supports IE anyways.

File: apps/web/src/app/settings/change-kdf.component.html [hunk 2]
```diff
@@ -32,41 +32,80 @@ <h1>{{ "encKeySettings" | i18n }}</h1>
         >
           <i class="bwi bwi-question-circle" aria-hidden="true"></i>
         </a>
-        <select id="kdf" name="Kdf" [(ngModel)]="kdf" class="form-control" required>
+        <select
```
[the remark this answers is not in the document]
Comment (quexten) [reply apps/web/src/app/settings/change-kdf.component.html:35]: Yeah.

File: apps/web/src/app/settings/change-kdf.component.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (quexten) [reply apps/web/src/app/settings/change-kdf.component.html]: Removed in 2a937d0fa7ff8b6ec94c3957ae882604142bcb23

File: apps/web/src/app/settings/change-kdf.component.html [hunk 3]
```diff
@@ -39,34 +39,66 @@ <h1>{{ "encKeySettings" | i18n }}</h1>
...
-          id="kdfIterations"
-          type="number"
-          min="5000"
-          max="2000000"
-          name="KdfIterations"
-          class="form-control"
-          [(ngModel)]="kdfIterations"
-          required
-        />
+        <ng-container *ngIf="kdf == 0">
```
[the remark this answers is not in the document]
Comment (quexten) [reply apps/web/src/app/settings/change-kdf.component.html]: Fixed in e3a5808ac107fcf508860f76e6de50d08cb54858

**Comment by Hinton** [comment]
@kspearrin @quexten there is webpack flag, `asyncWebAssembly` https://webpack.js.org/configuration/experiments/ which should let you remove most of the webpack changes.

**Comment by quexten** [comment]
> @kspearrin @quexten there is webpack flag, `asyncWebAssembly` https://webpack.js.org/configuration/experiments/ which should let you remove all the webpack changes. (At least that's how I got webassembly working in the web vault a while back). Would be nice figuring out why we need the csp change as well.

I did see this flag during development, but didn't use it as it was experimental. But it seems it is enabled under future defaults? I'll switch to it at any rate.

**Comment by quexten** [comment]
> @kspearrin @quexten there is webpack flag, `asyncWebAssembly` https://webpack.js.org/configuration/experiments/ which should let you remove all the webpack changes. (At least that's how I got webassembly working in the web vault a while back). Would be nice figuring out why we need the csp change as well.

Hmm, so I tried the flag and I was able to remove the line:
`    noParse: /\.wasm$/,`

However, still required is:
```
  {
    test: /\.wasm$/,
    loader: "base64-loader",
    type: "javascript/auto",
  },
```
because of:
```
ERROR in ../../node_modules/argon2-browser/dist/argon2.wasm 
Module not found: Error: Can't resolve 'a' in '/home/quexten/projects/clients/node_modules/argon2-browser/dist'
```
(not exactly sure yet what the exact cause is)



About the CSP changes:
removing wasm-eval should be fine as wasm-unsafe-eval is preferred these days, but it seems chrome still requires this in extensions?

Ideally we could use SRI hashes, but for WASM, these are unsupported in chromium.

**Comment by Hinton** [comment]
I toyed around with the wasm stuff a bit further. This wasm dependency seems to work quite different from the rust one I built myself. It seems we do need the webpack changes, and the `experimentals.asyncWebAssembly` doesn't do anything for us.

Do we have any warnings in case the wasm execution fails due to `wasm-eval` or `wasm-unsafe-eval` isn't supported? We would want to block the user from logging in + changing the KDF with a nice message vs silently logging errors in the console.

File: libs/node/package.json [hunk 4]
```diff
@@ -18,6 +18,7 @@
     "build:watch": "npm run clean && tsc -watch"
   },
   "dependencies": {
-    "@bitwarden/common": "file:../common"
+    "@bitwarden/common": "file:../common",
+    "argon2": "^0.30.3"
```
[the remark this answers is not in the document]
Comment (Hinton) [reply libs/node/package.json:22]: No, packages should only be in the root package.json. All of these should be moved over.

File: libs/common/package.json [hunk 5]
```diff
@@ -16,5 +16,9 @@
     "clean": "rimraf dist/**/*",
     "build": "npm run clean && tsc",
     "build:watch": "npm run clean && tsc -watch"
+  },
+  "dependencies": {
```
Comment (Hinton) [inline libs/common/package.json:20]: These should be moved to the root package.json. `base64-loader´ should be a dev dependency. And please include the updates to the package-lock files.

**Comment by kspearrin** [comment]
@quexten I finished updating the server PR with the necessary changes. I think @justindbaur is adding some tests to it.

Did you start a branch somewhere else with the separate config changes yet?

File: apps/web/src/app/settings/change-kdf.component.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtigas) [reply apps/web/src/app/settings/change-kdf.component.html]: Unsure if there are other configurations out there like this (i.e., first-party supported by a major OS/browser vendor), but

- enabling [Lockdown Mode](https://support.apple.com/en-us/HT212650) on iOS disables WASM (among other browser features) on Safari and any WebView implementation in all apps (including other browser apps).
  - Attempting to access the web vault via Safari or any other browser is affected.
  - I don't believe the Bitwarden native iOS app is affected (AFAIK the only WebView is for webauthn)
- Lockdown Mode is also available on macOS, but AFAIK only Safari is affected.
  - The Bitwarden browser extension within Safari _is_ affected, and I have not found a way to whitelist it. (The UI svg/webfont icons for "copy username", "copy password", etc, do not render and are simply shown as squares. There's support for whitelisting domains from Lockdown Mode, but the `safari-web-extension://${UUID}` URIs can't seem to be whitelisted.)

**Comment by quexten** [comment]
> Did you start a branch somewhere else with the separate config changes yet?

Not yet, I have the old branch, which I need to re-base onto this. I'll do that now.

**Comment by quexten** [comment]
> @quexten I finished updating the server PR with the necessary changes. I think @justindbaur is adding some tests to it.
> 
> Did you start a branch somewhere else with the separate config changes yet?

https://github.com/quexten/clients/pull/1

I targeted it to this pull request's source branch for now. Let me know whether I should merge it into this PR's source or submit it separately against bitwarden's master branch.

**Comment by kspearrin** [comment]
@quexten Let's finish cleaning up this branch, merge it, and then we can target master and open a PR for your new config branch.

File: libs/node/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (quexten) [reply libs/node/package.json:22]: Anything we need to consider for packaging for the different app stores? I only tested running in dev mode.

[the remark this answers is not in the document]
Comment (kspearrin) [reply libs/node/package.json]: I moved these to root.

**Comment by kspearrin** [comment]
@quexten I did some work cleaning things up here. I think it is good to go after we remove the changes to web angular components and locale files.

**Comment by micahblut** [comment]
@quexten we will create a help document for users comparing the two algorithms. Can you update the kdf algorithms tooltip link to point to https://bitwarden.com/help/kdf-algorithms

**Comment by kspearrin** [comment]
@micahblut We can do that in the configuration branch.

**Comment by danielleflinn** [comment]
Feedback from Bitwarden Design:

- Update the form so that there is not a large gap on the left side when Argon2id is selected
- Use sentence case on field labels
- Helper text: "Higher Argon 2 iterations, memory, and parallelism..." (update order to match field order and use the oxford comma)
- In Callout, update sentence to "...WebAssembly, you cannot change your KDF algorithm to Argon2." Remove the “}” that is appearing at the end of the callout’s text in the previously posted screenshot
- If after submit we know the submission failed from lack of WebAssembly support, show an error toast notification with the following: "Browser does not support WebAssembly. Try again with a different browser"
- Update the PBKDF2 encryption text based warning to use the warning callout. This provider better UI consistency between the 2 encryption options. 

image

File: apps/browser/webpack.config.js [hunk 6]
```diff
@@ -203,13 +208,18 @@ const mainConfig = {
       buffer: require.resolve("buffer/"),
       util: require.resolve("util/"),
       url: require.resolve("url/"),
+      fs: false,
+      path: false,
```
Comment (kspearrin) [inline apps/browser/webpack.config.js:212]: @quexten Can you comment why these fallbacks were added?

File: apps/web/webpack.config.js [hunk 7]
```diff
@@ -223,6 +228,8 @@ const devServer =
                 default-src 'self'
                 ;script-src
                   'self'
+                  'wasm-eval'
+                  'wasm-unsafe-eval'
```
Comment (kspearrin) [inline apps/web/webpack.config.js:232]: Are both of these required?

File: apps/web/webpack.config.js [hunk 8]
```diff
@@ -349,13 +356,19 @@ const webpackConfig = {
       util: require.resolve("util/"),
       assert: false,
       url: false,
+      path: false,
+      fs: false,
+      process: false,
```
Comment (kspearrin) [inline apps/web/webpack.config.js:361]: Same question. Why are these needed?

Comment (quexten) [reply apps/web/webpack.config.js:361]: Building without them results in:

```
ERROR in ../../node_modules/argon2-browser/dist/argon2.js 43:22-45
Module not found: Error: Can't resolve 'path' in '/home/quexten/tmp/clients/node_modules/argon2-browser/dist'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "path": require.resolve("path-browserify") }'
	- install 'path-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "path": false }
resolve 'path' in '/home/quexten/tmp/clients/node_modules/argon2-browser/dist'
  Parsed request is a module
  using description file: /home/quexten/tmp/clients/node_modules/argon2-browser/package.json (relative path: ./dist)
    Field 'browser' doesn't contain a valid alias configuration
    resolve as module
      looking for modules in /home/quexten/tmp/clients/node_modules
        single file module
          using description file: /home/quexten/tmp/clients/package.json (relative path: ./node_modules/path)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              /home/quexten/tmp/clients/node_modules/path doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              /home/quexten/tmp/clients/node_modules/path.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              /home/quexten/tmp/clients/node_modules/path.js doesn't exist
        /home/quexten/tmp/clients/node_modules/path doesn't exist
 @ ../../node_modules/argon2-browser/lib/argon2.js 72:29-57 75:11-38
 @ ../../libs/common/src/services/webCryptoFunction.service.ts 10:0-41 39:37-48 46:22-47
 @ ../../libs/angular/src/services/jslib-services.module.ts 112:0-96 415:22-46
 @ ./src/app/core/core.module.ts 4:0-88 105:31-50
 @ ./src/app/core/index.ts 1:0-30 1:0-30
 @ ./src/app/app.module.ts 7:0-36 19:8-18
 @ ./src/main.ts 8:0-45 12:54-63

ERROR in ../../node_modules/argon2-browser/dist/argon2.js 49:26-39
Module not found: Error: Can't resolve 'fs' in '/home/quexten/tmp/clients/node_modules/argon2-browser/dist'
resolve 'fs' in '/home/quexten/tmp/clients/node_modules/argon2-browser/dist'
  Parsed request is a module
  using description file: /home/quexten/tmp/clients/node_modules/argon2-browser/package.json (relative path: ./dist)
    Field 'browser' doesn't contain a valid alias configuration
    resolve as module
      looking for modules in /home/quexten/tmp/clients/node_modules
        single file module
          using description file: /home/quexten/tmp/clients/package.json (relative path: ./node_modules/fs)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              /home/quexten/tmp/clients/node_modules/fs doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              /home/quexten/tmp/clients/node_modules/fs.ts doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              /home/quexten/tmp/clients/node_modules/fs.js doesn't exist
        /home/quexten/tmp/clients/node_modules/fs doesn't exist
 @ ../../node_modules/argon2-browser/lib/argon2.js 72:29-57 75:11-38
 @ ../../libs/common/src/services/webCryptoFunction.service.ts 10:0-41 39:37-48 46:22-47
 @ ../../libs/angular/src/services/jslib-services.module.ts 112:0-96 415:22-46
 @ ./src/app/core/core.module.ts 4:0-88 105:31-50
 @ ./src/app/core/index.ts 1:0-30 1:0-30
 @ ./src/app/app.module.ts 7:0-36 19:8-18
 @ ./src/main.ts 8:0-45 12:54-63
```

Due to the minified argon2-browser code including requires for path, fs and process.

File: apps/web/webpack.config.js (hunk 7 above)
Comment (quexten) [reply apps/web/webpack.config.js:232]: From what I have read, `wasm-eval` is the preliminary version and should be safe to remove, since all browsers *should* implement wasm-unsafe-eval now. But I haven't tested this on all platforms (browser extensions).

**Review by kspearrin — APPROVED** [review]

File: libs/common/src/enums/kdfType.ts [hunk 9]
```diff
@@ -1,7 +1,12 @@
 export enum KdfType {
   PBKDF2_SHA256 = 0,
+  Argon2id = 1,
 }
 
+export const DEFAULT_ARGON2_MEMORY = 16;
+export const DEFAULT_ARGON2_PARALLELISM = 2;
+export const DEFAULT_ARGON2_ITERATIONS = 2;
```
Comment (mweinelt) [inline libs/common/src/enums/kdfType.ts:8]: Should probably be updated to account for the latest changes here: https://github.com/OWASP/CheatSheetSeries/commit/f14bc4c0578e1328c23094dcb0a1d8e9b778b3ec

```suggestion
export const DEFAULT_ARGON2_MEMORY = 19;
export const DEFAULT_ARGON2_PARALLELISM = 2;
export const DEFAULT_ARGON2_ITERATIONS = 2;
```

**Review by Hinton — COMMENTED** [review]
@kspearrin maybe I'm missing something, but there doesn't seem to be any way to configure the argon2 options?

We should also track and show a reasonable error if we can't load the argon2 wasm and block the login on those devices / changing kdf.

Other than that I think we're good. Do we want a feature flag for this until everything is wrapped up?

**Comment by quexten** [comment]
> @kspearrin maybe I'm missing something, but there doesn't seem to be any way to configure the argon2 options?
> 
> We should also track and show a reasonable error if we can't load the argon2 wasm and block the login on those devices / changing kdf.
> 
> Other than that I think we're good. Do we want a feature flag for this until everything is wrapped up?

Config options got rolled back in 9537c5ce589df8899cb0009b534f08e93b2c3e44.

So at the moment, the clients can log in if the account is configured for argon2 but there is no way to switch to argon2, or configure argon2's parameters.

Separate pull request for configuration that will be re-targeted to master once this is merged is here:
https://github.com/quexten/clients/pull/1

**Comment by kspearrin** [comment]
@Hinton I don't think we need a feature flag since everything is planned to be in this release.

Config options will follow after this is merged.

**Review by kspearrin — APPROVED** [review]

**Review by Hinton — APPROVED** [review]

**Comment by kspearrin** [comment]
@quexten Please open your config PR when ready.

**Comment by kspearrin** [comment]
@quexten I can help with that. Will move to #4578
